react/AppIcon/index.jsx

```diff
@@ -0,0 +1,88 @@
+import React, { Component } from 'react'
+import PropTypes from 'prop-types'
+import classNames from 'classnames'
+import styles from './styles.styl'
+
+import Icon from '../Icon'
+
+import appDefaultIcon from '../../assets/icons/ui/icon-cube.svg'
+
+export class AppIcon extends Component {
+  constructor(props, context) {
+    super(props, context)
+    this.state = {
+      status: 'idle'
+    }
+  }
+
+  async componentDidMount() {
+    const { app } = this.props
+    const { client } = this.context
+    const cozyClient = client && client.link && client.link.client
+    if (!cozyClient || typeof cozyClient.fetch !== 'function') {
+      console.warn(
+        'AppIcon needs CozyClient and to be called into a <CozyClientProvider /> component.'
+      )
+      this.setState({ status: 'failed' })
+      return
+    }
+    const icon = await this.fetchIcon(app.links.icon, cozyClient)
+    this.setState({ icon, status: icon ? 'done' : 'failed' })
+  }
+
+  async fetchIcon(url, cozyClient) {
+    this.setState({ status: 'fetching' })
+    let icon
+    try {
+      const resp = await cozyClient.fetch('GET', url)
+      if (!resp.ok)
+        throw new Error(`Error while fetching icon ${resp.statusText}: ${url}`)
+      icon = await resp.blob()
+    } catch (error) {
+      console.warn(error.message)
+      return
+    }
+    // check if MIME type is an image
+    if (!icon.type.match(/^image\/.*$/)) {
+      console.warn(`App icon ${url} is not an image.`)
+      return
+    }
+    return URL.createObjectURL(icon)
+  }
+
+  render() {
+    const { alt } = this.props
+    const { icon, status } = this.state
+    switch (status) {
+      case 'idle':
+      case 'fetching':
+        return (
+          <div
+            className={classNames(
+              styles['c-loading-placeholder'],
+              styles['c-app-icon'],
+              styles['c-app-icon-placeholder']
```

Comment:
I wonder if these 3 classes could be replaced by only 1 ? What happens if one is removed ? Do they need to be used together or have they orthogonal concerns ?

Reply:
Actually, `c-app-icon` is used for all icon "states": `done`, `fetching` and `failed`, even if its content change. Both `(.)*-placeholder` classes are used for fetching animation. In the store all the loading animations have a common basis, which has been used here for `c-loading-placeholder`, and a specific `animation-delay` for different elements. This specific part has been kept here into `c-app-icon-placeholder`, but you're right we could actually merge both of theme.

Loading animation need to be integrated into Cozy UI, but it's not a simple work. When it will be done we will update this part.

For now, I merge those two `.*-placeholder` classes.